Update private_key.md

youngminpark2559 opened the pull request:
For not making readers confused by giving them the idea that Bitcoin secret and private key are separated things.

youngminpark2559 committed:
Update private_key.md

For not making readers confused by giving them the idea that Bitcoin secret and private key are separated things.

NicolasDorier commented:
I agree, the turn of the sentence is a bit weird though. 
I would have said
```
Bitcoin secret is nothing more than Private key represented in Base58Check
```

youngminpark2559 committed:
Update comment

By getting Nicolas advice.

youngmtool commented:
Sir, how do you think that below content will be attached in "private key" section as a Tip?
There are some wrong idea? Or is this needless? 
In Bitcoin, we don't put raw private key into wallet or transaction for security reason.
Instead, we hash and encode raw private key into some format such as Bitcoin Secret represented in Base58Check, and generally use private key in such format. But again, Bitcoin secret is nothing more than Private key represented in Base58Check.
Since this format of private key can be used with Wallet, Bitcoin Secret is also known as Wallet Import Format(WIF).
You can see the more details and uses in "spend your coins" section of this chapter.

youngminpark2559 committed:
Is adding "key" along with address more precise?

I saw like that in BIP38 (Part 2) section.

youngmtool commented:
Please anyone tells me if it's something wrong or not.
In P2PK[H] (Pay to Public Key [Hash]) section, the sentence says We learned that a Bitcoin Address was the hash of a public key.
However, they're different like this:
Console.WriteLine(publicKeyHash); // 41e0d7ab8af1ba5452b824116a31357dc931cf28
Console.WriteLine(bitcoinAddress); // 171LGoEKyVzgQstGwnTHVh3TFTgo5PsqiY

NicolasDorier commented:
Using Base58 does not increase security. It is just a standard to present and import public key in a human friendly way.

Code should normally use `Key`, and only use `BitcoinSecret` at the UI Layer.

```
But for preventing from being confused note that Bitcoin secret is nothing more than Private key represented in Base58Check.
```

This sentence is strange. Maybe should you add some punctuation or remove "But for preventing from being confused".

NicolasDorier commented:
Bitcoin Address is the hash of the public key encoded in Base58.

youngminpark2559 committed:
Update private_key.md

Thank for explanations.
I personally think that in that context, "But for preventing from being confused" is better placed there for understanding. So, I put a punctuation at the end of it.

youngminpark2559 committed:
Add contents about CombineSignatures()

youngminpark2559 committed:
Merge pull request #1 from youngmtool/youngmtool-patch-1

Add contents about CombineSignatures()

nopara73 committed:
Merge branch 'master' into master

nopara73 commented:
@bokobza I resolved this conflict the same way we resolved back then. As you can see it didn't get resolved, just like back then. Do you remember what was the GitHub bug we encountered and how we solved it?

youngminpark2559 committed:
Update bitcoin_address.md

I was confused when I was first reading this section by Key privateKey = new Key(); // generate a random private key
new Key() can be said as a private key? I think it's more near to a kind of key object.
So I edit that, if it's not appropriate, I will back to the original one.
And I add some contents for giving background basics about Base64, Base58 etc.

youngminpark2559 committed:
Edit code block separation

youngminpark2559 committed:
Fix grammatical typos and add section separations

NicolasDorier commented on other_types_of_ownership/multi_sig.md:
```diff
@@ -116,6 +118,44 @@ Console.WriteLine(fullySigned);
+Before sending the transaction to the network, examine the need of CombineSignatures() method.
+Try to compare a full detail of transaction between bobSigned and fullySigned.
+It will seem both are identical.
+For this reason, it seems like the CombineSignatures() method is needless because mulit-signing has achieved without the CombineSignatures() method.
```
typo *mulit*

NicolasDorier commented on other_types_of_ownership/multi_sig.md:
```diff
@@ -116,6 +118,44 @@ Console.WriteLine(fullySigned);
+                    .BuildTransaction(sign: false);
+
+            
+            Transaction aliceSigned =
```
identation

NicolasDorier commented on other_types_of_ownership/multi_sig.md:
```diff
@@ -116,6 +118,44 @@ Console.WriteLine(fullySigned);
+                    .AddKeys(alice)
+                    .SignTransaction(unsignedNew);
+            
+            Transaction bobSigned =
```
identation

NicolasDorier commented on bitcoin_transfer/bitcoin_address.md:
```diff
@@ -7,52 +7,106 @@ You probably know that your wallet software uses a **private key** to spend the
 
-The Base58Check encoding has some neat features, such as checksums to prevent typos and a lack of ambiguous characters such as '0' and 'O'.  
-The Base58Check encoding also provides a consistent way to determine the network of a given address; preventing a wallet from sending MainNet coins to a TestNet address.
+## Encoindg schemes {# encoindg-schemes}
```
typo Encoindg

youngminpark2559 committed:
Fix multi-signing to multi-sig

youngminpark2559 committed:
Fix an indentation

youngminpark2559 committed:
Fix typo

youngmtool commented:
Thanks, Nicolas. I edited.
And I have a question.
In Multi Sig section, the words "multi-sig" and "ScriptPubKey" or the process of generating a ScriptPubKey don't make sense at the same time because sig means signing signature but generating a ScriptPubKey doesn't require signing signature but just a public key.
I think it's more appropriate that a ScriptPubKey is decorated by a kind of multi-publicKey rather than a multi-sig.

I wonder this.
Does the concept of multi-sig contain largely the following whole process?
1.Generating a ScriptPubKey by multiple publicKeys, and putting it into TxOut of an other getting coins transaction.
2.Proofing ownership of the recieved coins by multiple private keys or by multi-sig which is combined into one thing by multiple privatekeys.
3.Generating a scriptSig by processing multi-sig for a TxIn of a spending transaction.

youngminpark2559 committed:
Put a punctuation

youngminpark2559 committed:
Add additional explanation

youngmtool commented:
I also wonder this point.
1.
Can I solidify an idea that the process of generating a ScriptPubKey is eleminating a version byte which identifies a network type, and getting publicKey hash from a Bitcoin address(a version byte + a publicKey hash), and putting some script commands?

From above statement, can I consider the process of generating a Bitcoin address by a ScriptPubKey and network type as the same process of generating a Bitcoin address by a publicKey hash and network, except for direction(forwards for the latter, backwards for the former) and the fact that a ScriptPubKey only contains additional script commands, compared to a publicKey hash?

2.
Can a ScriptPubKey be said as an address? Or can it be said as a value?
Got wondered from this sentence:
Notice the ScriptPubKey for testnet and mainnet address is the same ?

youngminpark2559 committed:
Update bitcoin_address.md

youngminpark2559 committed:
Update bitcoin_address.md

youngminpark2559 committed:
Update bitcoin_address.md

youngmtool commented:
I wonder the need of a PrivateKey property in a BitcoinSecret class.

In this page: https://programmingblockchain.gitbooks.io/programmingblockchain/content/bitcoin_transfer/private_key.html

Around this sentence:
Note that it is easy to go from BitcoinSecret to private Key.

I'm not sure what meaning does the attempt of going from a BitcoinSecret type to a privateKey by specifying a PrivateKey property on a BitcoinSecret type has.
Because a PrivateKey property on a BitcoinSecret type returns a Key object, NBitcoin.Key.
If a privateKey property in the BitcoinSecret class returns only a Key object, I think there's no reason to exsist it in that class. Does that privateKey property has other roles? 

And also semantically of BitcoinSecret, WIF, and private key, under my understanding at the moment, they're just 3 different calling words for one concept, the private key.
So, it seems like tyring to go from BitcoinSecret to a private key doesn't make sense for me.

youngminpark2559 committed:
Add explanatory word

youngminpark2559 committed:
Update transaction.md

youngminpark2559 committed:
Update private_key.md

NicolasDorier commented:
> In Multi Sig section, the words "multi-sig" and "ScriptPubKey" or the process of generating a ScriptPubKey don't make sense at the same time because sig means signing signature but generating a ScriptPubKey doesn't require signing signature but just a public key.
I think it's more appropriate that a ScriptPubKey is decorated by a kind of multi-publicKey rather than a multi-sig.

No, a ScriptPubKey is not simply a public key. A ScriptPubKey is a script which fix the conditions to redeem the output. When you talk about native "multi-sig" it is a ScriptPubKey while looks like "2 <pub1> <pub2> OP_MULTISIG". In general, when we talk about "multi" we mean that we need several people to spend something.

1. and 2.:

Every Address represent a particular ScriptPubKey (OP_DUP OP_HASH160 <H> OP_EQUAL OP_CHECKSIG).

However all ScriptPubKey can't be represented as an address. 

> And also semantically of BitcoinSecret, WIF, and private key, under my understanding at the moment, they're just 3 different calling words for one concept, the private key.
So, it seems like tyring to go from BitcoinSecret to a private key doesn't make sense for me.

BitcoinSecret denotes a Base58 string representation of the key (it has network information).
Your processing code should only use Key, and BitcoinSecret should only be used as UI. (Exporting/Importing/Saving a private key)

youngmtool commented:
Thank you. Helpful for my lack basics.
But on the 3rd one, 

you said, BitcoinSecret denotes a Base58, but is a BitcoinSecret represented in a Base58Check?
I saw like that in Privete key section in this book and also saw a BitcoinSecret constructor which requires a Base58 type and a network type for arguments.
After newly encoded with those informations and checksum data, it finally becomes a Base58Check which is different from a Base58?

The Key you mentioned here sentence:
Your processing code should only use Key.
means a key object(new Key())?
If I should use a Key you mentioned only in code and should use a BitcoinSecret(=WIF) in UI such as wallet software for Exporting/Importing/Saving a private key, where is a Key you mentioned?
Is it a random key object itself in which I only can see various members?
Moreover, when I examine a Key class, there are only GetBitcoinSecret(), GetWif() which are related to a private key.
(And can I say that BitcoinSecret and Wif are exactly the same?)

And you gave a Key object(new Key()) a name, privateKey, is it wrong idea if I call a new Key() a keyObject?

NicolasDorier commented:
not sure what you mean. Yes BitcoinSecret is wif/base58check data.

`Key` is the class you should use internally for any operations on private keys.

youngminpark2559 committed:
Update spend_your_coin.md

youngminpark2559 committed:
Update spend_your_coin.md

youngminpark2559 committed:
Update spend_your_coin.md

youngminpark2559 committed:
Please anyone check if it is how

IMHO, I think we're not using QBitNinja transactionResponse there.

youngminpark2559 committed:
Update spend_your_coin.md

youngmtool commented:
Well, first of all, please examine a updated section by me:
https://github.com/youngmtool/ProgrammingBlockchain/blob/master/bitcoin_transfer/bitcoin_address.md
And this is the original not updated:
https://github.com/ProgrammingBlockchain/ProgrammingBlockchain/blob/master/bitcoin_transfer/bitcoin_address.md
And the corresponding commit:
https://github.com/youngmtool/ProgrammingBlockchain/commit/7b44a07e49e40674100d6a901c4a42c279a17b56
Please see the upper part, describing keyObj, keyObj.PubKey...

And of things related to above statement in the current comment, about private key, BitcoinSecret, reference comment again what I wrote, this one:
I wonder the need of a PrivateKey property in a BitcoinSecret class.

And please exmaine the result of example:
```
            //Genarate a random key object.
            var keyObj1 = new Key();
            BitcoinSecret bitcoinSecretByGetWif = keyObj1.GetWif(Network.Main);
            BitcoinSecret bitcoinSecretByGetBitcoinSecret = keyObj1.GetBitcoinSecret(Network.Main);

            var privateKeyFronBitcoinSecret1 = bitcoinSecretByGetWif.PrivateKey;
            var privateKeyFronBitcoinSecret2 = bitcoinSecretByGetBitcoinSecret.PrivateKey;

            Console.WriteLine("===========");

            Console.WriteLine(bitcoinSecretByGetWif);
            Console.WriteLine(bitcoinSecretByGetBitcoinSecret);
            Console.WriteLine(keyObj1);
            Console.WriteLine(privateKeyFronBitcoinSecret1);
            Console.WriteLine(privateKeyFronBitcoinSecret2);
            //L3syD8Z3biLy9Qb2MfUGZNTe3yfFMYf9r8oRXVJPng4GM1My9jsv
            //L3syD8Z3biLy9Qb2MfUGZNTe3yfFMYf9r8oRXVJPng4GM1My9jsv
            //NBitcoin.Key
            //NBitcoin.Key
            //NBitcoin.Key
            //NBitcoin.Key is a private key you mentioned as a Key which should be used in code?
            //For me, it looks like a kind of a key object which has various member.
```
And examine again my above questions on a private key, a BitcoinSecrete, and so on.

youngminpark2559 committed:
Question.

Above one results in TestNet.
But at this page:
https://github.com/youngmtool/ProgrammingBlockchain/blob/master/bitcoin_transfer/proof_of_ownership_as_an_authentication_method.md
var bitcoinPrivateKey = new BitcoinSecret("XXXXXXXXXXXXXXXXXXXXXXXXXX");
I don't know why it's written by XXX, it's KzgjNRhcJ3HRjxVdFhv14BrYUKrYBzdoxQyR2iJBHG9SNGGgbmtC
See, example code:
https://github.com/youngmtool/ProgrammingBlockchainCodeExamples/blob/master/ProofOfOwnership/Program.cs
This one results in MainNet.

Without specifying a network, what makes differences in a network type?

And thing which is passed into BitcoinSecret constructor argument, KzgjNRhcJ3HRjxVdFhv14BrYUKrYBzdoxQyR2iJBHG9SNGGgbmtC, should be a private key because BitcoinSecret is generated by a private key and network.

So, how can I generate a private key like KzgjNRhcJ3HRjxVdFhv14BrYUKrYBzdoxQyR2iJBHG9SNGGgbmtC in NBitcoin?

According to Bitcoin address section of this book, if I do like this, Key k = new Key();
Console.WriteLine(k) doesn't show the one like KzgjNRhcJ3HRjxVdFhv14BrYUKrYBzdoxQyR2iJBHG9SNGGgbmtC.

youngminpark2559 committed:
Update key_generation.md

youngminpark2559 committed:
Update key_generation.md

youngmtool commented:
Changes are not merged yet.
https://github.com/ProgrammingBlockchain/ProgrammingBlockchain/blob/master/bitcoin_transfer/bitcoin_address.md

There was lack of explanations about this in the book.

youngmtool commented:
I understood a little bit after examining Key.cs, BitcoinSecret.cs, and Base58Data.cs from a NBitcoin repository.

But I have 2 questions.

1.
When I do this:
Key privateKey = new Key();
privateKey.ToString(Network.Main);

This instantiates BitcoinSecret class, passing privateKey object and network.

In BitcoinSecret class, it instantiates its base class, Base58Data, invoking ToBytes(Key key).

In this method, ToBytes() method for a Base58Data class is invoked on a key object.


In a Base58Data class, ToBytes() method returns vchData.ToArray().

But at this time, vchData is not assigned yet?

vchData is assigned by Byte[] data after invoking SetData() method in Base58Data constructor.

But the point confusing me is that Base58Data looks like it's instantiated via TyBytes() method first which returns vchData.ToArray().


2.
I read this:
When you call new Key(), under the hood, you are using a PRNG (Pseudo-Random-Number-Generator) to generate your private key.
from Is it random enough? section in this book.

But I don't see the places where a PRNG contributes to generatring a private key from the 3 class I examined.

And how a PRNG and a Key.cs is connected each other.
The connection is defined in kind of metadata in IL code?

NicolasDorier commented:
1.

I don't understand much. But this code is confusing. (I initially ported it straight from C++)
`vchData` is the raw bytes you get once you decode the Base58Check string.
In the case of BitcoinSecret, there is, if I remember, 1 bytes saying if the pubkey should be compressed or not, followed by 32 bytes of the private key.

2.

RNG called here https://github.com/MetacoSA/NBitcoin/blob/master/NBitcoin/Key.cs#L43

youngmtool commented:
Thank you so much for both answers.

youngminpark2559 committed:
Back some part(new Key(), privateKey) back to previous one in bitcoin_address.md

youngminpark2559 committed:
Update bitcoin_address.md

youngminpark2559 committed:
Update key_generation.md

youngminpark2559 committed:
Update key_generation.md

youngminpark2559 committed:
Add explanation of a key object

youngminpark2559 committed:
Update bitcoin_address.md

youngminpark2559 committed:
Update bitcoin_address.md

youngmtool commented:
Nicolas, I wonder these:

1.
I understood a key object contains a private key data generated by the RNGCryptoServiceProvider.

But what's the data type of the stored private key data in a key object?
Byte array?

2.
Is there any way how to print a key data on the console without being helped by BitcoinSecret constructor, when I do kind of privateKey.ToString(Network.Main);

3.
What's the meaning of using a Key in the code?

Does it mean that I just use a key object encoded or decoded between wallet software and codes interacting with BitcoinSecret which is a representation in Base58Check, without knowing a private key's specific raw data such as just by doing like this:
privateKey.GetWif(), privateKey.PubKey and so on.

And please review what I've added:
https://github.com/ProgrammingBlockchain/ProgrammingBlockchain/pull/81/commits/b13b5d7cf6dc18cc330f87d6e81c1bff8228fffd

youngminpark2559 committed:
Update key_generation.md

youngminpark2559 committed:
Update mnemonic-code-for-hd-keys-bip39

youngminpark2559 committed:
Update p2pk[h]_pay_to_public_key_[hash].md

youngminpark2559 committed:
Update arbitrary.md

youngminpark2559 committed:
Update arbitrary.md

youngminpark2559 committed:
Update arbitrary.md

youngminpark2559 committed:
Merge branch 'master' into master

youngminpark2559 committed:
Update private_key.md

youngminpark2559 committed:
Update private_key.md

youngminpark2559 committed:
Update bitcoin_address.md

youngminpark2559 committed:
Update README.md

youngminpark2559 committed:
Update bitcoin_address.md

NicolasDorier commented:
1. yes byte array of 32 bytes

2. `Encoders.Hex.EncodeData("key.ToBytes()")`

3. Not sure what you mean, a key is needed for many operations in Bitcoin. BitcoinSecret is a Key + Network. In Bitcoin processing code, Network information is never needed.

youngmtool commented:
Thanks the review. 
I tried this way.

1.
> I don't understand what you mean by Is it just a key object rather than specifying it as a private key

I thought this way,"Is it just a key object rather than specifying it as a private key?" or "Should it be named by keyObject rather than a privateKey?", when I was seeing this code:
Key privateKey = new Key();
So I thought readers would think this way too when they were seeing the code.
I'll change to the latter of the sentence which is more clear in the intention.

2.
> you instantiate a new object, not a class.

Fixed this way:
When you instantiate a Key object,

3.

> not necessarily true. On non-windows plateform it is untrue.

I think RNGCryptoServiceProvider things would be helpful for readers to understand widely, so I want to add explanations by your suggestion like this:
under the hood, in the case of .NET, you should invoke RNGCryptoServiceProvider to generate a key data and store it into a key object, which is a .NET wrapper around the Windows Cryto API. However, in the .NET, particular RNGCryptoServiceProvider is not mandatory, you can choose different generators you want to use. And also, since the RNGCryptoServiceProvider wouldn't work on the non-windows platform, you should choose appropriate generators on each platform.

youngminpark2559 committed:
Update bitcoin_address.md

youngminpark2559 committed:
Update arbitrary.md

youngminpark2559 committed:
Update arbitrary.md

NicolasDorier commented:
> I think RNGCryptoServiceProvider things would be helpful for readers to understand widely, so I want to add explanations by your suggestion like this:
under the hood, in the case of .NET, you should invoke RNGCryptoServiceProvider to generate a key data and store it into a key object, which is a .NET wrapper around the Windows Cryto API. However, in the .NET, particular RNGCryptoServiceProvider is not mandatory, you can choose different generators you want to use. And also, since the RNGCryptoServiceProvider wouldn't work on the non-windows platform, you should choose appropriate generators on each platform.

I prefer the book does not go into such details. People can easily look the source code, the only thing they need to know is that the default is using a secure RNG.

youngminpark2559 committed:
Update bitcoin_address.md

youngminpark2559 committed:
Update bitcoin_address.md

youngmtool commented:
Okay, I agree on it and I'll keep that point in mind when I'm writing.
Fixed:
https://github.com/ProgrammingBlockchain/ProgrammingBlockchain/pull/81/commits/42ce6edcbe6ff8fc11f66cb196f9d29a091cbc85
https://github.com/ProgrammingBlockchain/ProgrammingBlockchain/pull/81/commits/284c111853b18a03b605baa0fe2b7fe17464540b


I have some questions on the ScriptPubKey page:
https://github.com/ProgrammingBlockchain/ProgrammingBlockchain/blob/master/bitcoin_transfer/payment_script.md

1. 
Does that mean some ScriptPubKey doesn't contain the hashed public key like this?:
OP_DUP OP_HASH160 <Omitted hashed public key> OP_EQUALVERIFY OP_CHECKSIG

2.
I understood like this.
ScriptPubKey is composed of both of script commands which are opcodes and hashed public key.
Script commands may play a role explaining the conditions which should be met to claim ownership of the coins.
Hashed public key related to corresponding public key and private key may play a role explaining the address for the recipient.

But this sentence:
A ScriptPubKey does not necessarily contain the hashed public key(s) permitted to spend the bitcoin.

Hashed public key is the thing permitting someone to spend the coin which is the task related to proof the ownership of the coin?

Under my understanding, I think hashed public key just represents the recipient like the address, not being related to some tasks of permitting coins.


3.
ScriptPubKey seems like it doesn't have the network information because hashed public key doesn't have it.
Blockchain sees the ScriptPubKey as the address of recipient but how does Blockchain distinguish the kind of network?

youngminpark2559 committed:
Update private_key.md

youngminpark2559 committed:
Update transaction.md

youngminpark2559 committed:
Update blockchain.md

youngminpark2559 committed:
Update blockchain.md

youngminpark2559 committed:
Update blockchain.md

youngminpark2559 committed:
Update blockchain.md

youngminpark2559 committed:
Update blockchain.md

youngminpark2559 committed:
Update blockchain.md

youngminpark2559 committed:
Update the_blockchain_is_more_than_just_bitcoin.md

youngminpark2559 committed:
Update proof_of_ownership_as_an_authentication_method.md

youngminpark2559 committed:
Update proof_of_ownership_as_an_authentication_method.md

youngminpark2559 committed:
Update spend_your_coin.md

youngminpark2559 committed:
Update colored_coins.md

NicolasDorier commented on bitcoin_transfer/blockchain.md:
```diff
@@ -4,8 +4,8 @@
 
-See the illustraion depicting a block and a blockchain, drawn by Youngmin Park, in the following link:
-
+See the illustraion depicting a block and a blockchain, drawn by Youngmin Park(Youngmtool), in the following link:
```
*illustraion typo, and a TxIn IS an input.

youngminpark2559 committed:
Update blockchain.md

youngmtool commented:
@NicolasDorier 
Thanks. Fixed. But of TxIn and input. That's the point I asked to @nopara73 
But I'm still confused by the terms for them.

Hope you review this thread from top one by one and give the opinion:
https://github.com/ProgrammingBlockchain/ProgrammingBlockchain/issues/94

And if you don't mind, I want to hear about above question of ScriptPubKey.

youngminpark2559 committed:
Update issuing_an_asset.md

youngminpark2559 committed:
Update README.md

youngminpark2559 committed:
Update bitcoin_address.md

youngminpark2559 committed:
Update payment_script.md

youngminpark2559 committed:
Update private_key.md
